### PR TITLE
[FLINK-14019][python][cli] Add command-line options for managing Python UDF environment and dependencies.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request adds command-line options for managing Python UDF environment and dependencies.*

NOTE: This feature is still [voting](http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/VOTE-Improve-the-Pyflink-command-line-options-Adjustment-to-FLIP-78-td35585.html), but we already reach an agreement on the [discussion](http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Improve-the-Pyflink-command-line-options-Adjustment-to-FLIP-78-td35440.html). In order to check in before code freezes, I open this PR so that interested people can start reviewing it.

## Brief change log

  - *Add options in `ClifrontendParser`*
  - *Add options in `PythonDriverOptionsParserFactory`*
  - *Pass options values from `PythonDriver` to python process*

## Verifying this change

This change is already covered by existing tests, such as *ClifrontendRunTests*, *test_python_option.py*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 1]
```diff
@@ -183,13 +213,19 @@
...
+		PY_OPTION.setArgName("pythonFile");
 
 		PYFILES_OPTION.setRequired(false);
-		PYFILES_OPTION.setArgName("pyFiles");
+		PYFILES_OPTION.setArgName("pythonFiles");
 
 		PYMODULE_OPTION.setRequired(false);
 		PYMODULE_OPTION.setArgName("pyModule");
+
+		PYREQUIREMENTS_OPTION.setRequired(false);
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: The reason of not setting the argument name of these options is that they are too long and take up too much space on the print result of "./flink run -h", which makes it very ugly.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 2]
```diff
@@ -183,13 +213,19 @@
...
+
+		PYARCHIVE_OPTION.setRequired(false);
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: ditto

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 3]
```diff
@@ -183,13 +213,19 @@
...
+
+		PYEXEC_OPTION.setRequired(false);
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: ditto

File: flink-python/pyflink/common/tests/test_python_option.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+             "python_file_1_{uuid}": "hdfs://file2.zip",
+             "python_file_2_{uuid}": "file3.egg",
+             "python_archive_3_{uuid}": "/py3.zip",
+             "python_archive_4_{uuid}": "/py3.zip",
+             "python_archive_5_{uuid}": "data.zip",
+             "python_requirements_file_6_{uuid}": "a.txt",
+             "python_requirements_cache_7_{uuid}": "b_dir"}, registered_files)
+
+    def test_python_options_integrated(self):
+        tmp_dir = tempfile.mkdtemp(dir=self.tempdir)
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/pyflink/common/tests/test_python_option.py]: The tmp_dir is used as a python module directory in this test case and does not used to store anything. Its name is misleading. I have rename these variables to more reasonable names.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have updated the PR. Please take a look again. :)

**Comment by sunjincheng121** [comment]
Hi, @aljoscha do you want have the final review of this PR? :)

File: flink-python/pyflink/common/tests/test_python_option.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class PythonOptionTests(PyFlinkTestCase):
+
+    def setUp(self):
+        self.j_env = TestEnv()
+        self.config = Configuration()
+        self.dependency_manager = DependencyManager(self.config, self.j_env)
+
+    def test_python_options(self):
```
Comment (dianfu) [inline flink-python/pyflink/common/tests/test_python_option.py]: What about renaming to test_load_from_env some other name like this?

File: flink-python/pyflink/common/tests/test_python_option.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            {"python_file_0_{uuid}": "/file1.py",
+             "python_file_1_{uuid}": "hdfs://file2.zip",
+             "python_file_2_{uuid}": "file3.egg",
+             "python_archive_3_{uuid}": "/py3.zip",
+             "python_archive_4_{uuid}": "/py3.zip",
+             "python_archive_5_{uuid}": "data.zip",
+             "python_requirements_file_6_{uuid}": "a.txt",
+             "python_requirements_cache_7_{uuid}": "b_dir"}, registered_files)
+
+    def test_python_options_integrated(self):
```
Comment (dianfu) [inline flink-python/pyflink/common/tests/test_python_option.py]: What about removing this test as I think we can add Python integration test in the flink-end-to-end-tests in the future? Personally I think this test is a little complicate and it's difficult to maintain.

File: flink-python/pyflink/testing/python_option_verifier.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+################################################################################
+import argparse
+import json
+import os
+
+from pyflink.common.dependency_manager import DependencyManager
+from pyflink.testing.test_case_utils import (TestEnv, TestConfiguration, replace_uuid,
+                                             decode_from_base64, sort_dict_by_key)
+
+if __name__ == "__main__":
```
Comment (dianfu) [inline flink-python/pyflink/testing/python_option_verifier.py]: Can remove this file if test_python_options_integrated is removed.

File: flink-python/pyflink/util/utils.py [hunk 8]
```diff
@@ -21,14 +21,15 @@
 from pyflink.java_gateway import get_gateway
 
 
-def to_jarray(j_type, arr):
+def to_jarray(j_type, arr, gateway=None):
```
Comment (dianfu) [inline flink-python/pyflink/util/utils.py]: This change can be reverted if test_python_options_integrated is removed.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review again! I have addressed your comments in the latest commit.

**Comment by dianfu** [comment]
@WeiZhong94 Thanks a lot for the update. LGTM.

**Comment by hequn8128** [comment]
Thank you all for the PR and review. Merging...

**Pull request closed without merging**